File: lang_python/oif_impl/openinterfaces/_impl/qeq/py_qeq_solver/qeq_solver.py
import math

import numpy as np
from openinterfaces._impl.interfaces.qeq import QeqInterface


class QeqSolver(QeqInterface):
    def solve_qeq(self, a: float, b: float, c: float, roots: np.ndarray) -> int:
        """Solve quadratic equation ax**2 + bx + c = 0.

        Assumes that the output array `roots` always has two elements,
        so if the roots are repeated, they both will be still present.

        Parameters
        ----------
        a, b, c : float
            Coefficients.
        roots: np.ndarray[2]
            Array to which both roots are written.

        Returns
        -------
        int
            Status code.
        """
        if a == 0.0:
            roots[0] = -c / b
            roots[1] = -c / b
        else:
            D = b**2 - 4 * a * c
            if b > 0:
                roots[0] = (-b - math.sqrt(D)) / (2 * a)
                roots[1] = c / (a * roots[0])
            else:
                roots[0] = (-b + math.sqrt(D)) / (2 * a)
                roots[1] = c / (a * roots[0])

        return 0

File: lang_c/oif_impl/_impl/qeq/jl_qeq_solver/qeq_solver.jl
module QeqSolver
export solve_qeq

# This solver does not actually require any context, so Self is empty.
mutable struct Self end

function solve_qeq(self::Self, a::Float64, b::Float64, c::Float64, roots::Vector{Float64})
    println("Solving quadratic equation with a=$a, b=$b, c=$c")
    if a == 0
        roots[1] = -c / b
        roots[2] = -c / b
    else
        D = b^2 - 4 * a * c
        if b > 0
            roots[1] = (-b - sqrt(D)) / (2 * a)
            roots[2] = c / (a * roots[1])
        else
            roots[1] = (-b + sqrt(D)) / (2 * a)
            roots[2] = c / (a * roots[1])
        end
    end

    return 0
end
end
